source/common/filesystem/io_file_error.h

```diff
@@ -0,0 +1,40 @@
+#pragma once
+
+#include <string>
+
+#include "envoy/api/io_error.h"
+
+#include "common/common/assert.h"
+
+namespace Envoy {
+namespace Filesystem {
+
+class IoFileError : public Api::IoError {
+public:
+  explicit IoFileError(int sys_errno) : errno_(sys_errno) {}
+
+  ~IoFileError() override {}
+
+private:
+  IoErrorCode errorCode() const override;
+
+  std::string errorDetails() const override;
+
+  int errno_;
+};
+
+using IoFileErrorPtr = std::unique_ptr<IoFileError, Api::IoErrorDeleterType>;
+
+template <typename T> Api::IoCallResult<T> resultFailure(T result, int sys_errno) {
+  return {result, IoFileErrorPtr(new IoFileError(sys_errno), [](Api::IoError* err) {
+            ASSERT(err != nullptr);
+            delete err;
```

Comment:
Bare news and deletes are always a bit concerning. Can you elaborate on why they are needed here?

Reply:
The implementations of Api::IoError need to use a custom deleter, see https://github.com/envoyproxy/envoy/blob/caf7ab123964cedd172a2d4cb29b2f2e05ca9156/source/common/network/io_socket_handle_impl.cc#L42 for example

Reply:
I see why this is necessary because of the trick that I think is being removed as part of #6037. I'm a little confused here because you're not special casing the deleter for `ENVOY_ERROR_AGAIN`, does it not need to be covered here? After #6037, maybe the custom deleter requirement can be removed?

Reply:
It looks like we don't need to cover the `ENVOY_ERROR_AGAIN` in this case.
Probably after the #6037 PR will be merged, some changes will need to be done in this code.